src/EnergyPlus/HeatRecovery.cc

```diff
@@ -2612,7 +2612,7 @@ namespace HeatRecovery {
                 HXPartLoadRatio = max(0.0, HXPartLoadRatio);
                 HXPartLoadRatio = min(1.0, HXPartLoadRatio);
 
-            } else if (CompanionCoilIndex > 0) {
+            } else if (CompanionCoilIndex > 0 && !state.dataDXCoils->DXCoilPartLoadRatio.empty()) {
```

Comment:
And a crash ocurred here during issue investigation. An alternate approach would be to ensure the DX coil were read in prior to calling HeatRecovery. I chose to protect here instead of executing another call to an object's getInput (since that happened a lot in this file and caused the issue).

Reply:
The logic of this if/else if block seems flawed.

    if (!state.dataHeatRecovery->CalledFromParentObject) {
    } else if (CompanionCoilIndex > 0) {
    }

The SimHeatRecovery determines how CalledFromParentObject is set using:

        if (present(HXUnitEnable)) {
            HXUnitOn = HXUnitEnable;
            //   When state.dataHeatRecovery->CalledFromParentObject is TRUE, this SIM routine was called by a parent object that passed in
            //   HXUnitEnable. HX will use the DX coil part-load ratio (optional CompanionCoilIndex must be present) or PLR passed in if not used with
            //   DX coil (optional CompanionCoilIndex must not be present).
            state.dataHeatRecovery->CalledFromParentObject = true;
        } else {
            //   HX is placed on a BRANCH, optional arguments are not passed in from SimAirServingZones.
            //   HX will calculate its own part-load ratio if optional HXUnitEnable flag is not present
            HXUnitOn = true;
            state.dataHeatRecovery->CalledFromParentObject = false;
        }

And if this DX coil is inside a HXAssisted object, that optional argument is always passed:

            HeatRecovery::SimHeatRecovery(state,
                                          thisHXCoil.HeatExchangerName,
                                          FirstHVACIteration,
                                          thisHXCoil.HeatExchangerIndex,
                                          FanOpMode,
                                          PartLoadRatio,
                                          HXUnitOn,

Which means the else if will be used at line 2615 above. This also means HXPartLoadRatio is determined by the coil object (line 2617 below). Except that for VS DX coils, the DXCoilPartLoadRatio array is not used. So since VS DX coils are allowed in the DX HXAssisted object, and CalledFromParentObject is always true for this configuration (i.e., the `else if` is hit), the model is either using the wrong PLR (assuming other non-VS DX coil exist), or there will be a crash. So I think the crash here revealed something much deeper than just array allocation timing.

    CoilSystem:Cooling:DX:HeatExchangerAssisted,
      A4 , \field Cooling Coil Object Type
           \required-field
           \type choice
           \key Coil:Cooling:DX
           \key Coil:Cooling:DX:SingleSpeed
           \key Coil:Cooling:DX:VariableSpeed